### PR TITLE
Adds the `fullscreen_mode` property to Node

### LINKED ISSUE #1: [Request] Get an `is_fullscreen` property for container/node.
Hi,

really like the library so far. Nicer for me to use than the python version.
But there's one feature that I'm really missing, and that is knowing if a container/node is currently fullscreen or not. This one can be optional if i3 does not always provide this, but it would be very helpful to be able to get the value.

### PR DESCRIPTION
**Pull request by ChristianJacobsen**
I decided to fix this issue myself :)
Please tell me if you want the data type changed or the mode changed into a class enum.

Fixes #1

### CODE REVIEW

**Comment by Iskustvo** [comment]
Hi @ChristianJacobsen,

First of all, I am very glad that you decided to use my library, thank you for that!

I appreciate the PR, however I found it very odd that `i3` didn't document this useful info even in their latest documentation so I've tried you changes locally and added printout in [get_tree.cpp](https://github.com/Iskustvo/i3-ipcpp/blob/master/examples/get_tree.cpp) example.

On my very simple setup, there were 3 nodes/containers that claimed they were in fullscreen mode.

1. First one was some kind of internal `i3` workspace named `__i3_scratch`.
2. Second one was my (second) workspace which actually contained the node in fullscreen.
3. Third one was the node(terminal) which I've placed in fullscreen.

This was a bit strange since I expected only **3.** to be registered as fullscreen node.
I thought that maybe all parents of the node are also noted as fullscreen, but further testing showed that even that assumption was wrong.

I don't see the clear logic for this `fullscreen_mode` info to be useful.
It feels like registering to the `window_event` and checking if change was regarding the fullscreen status is way better/easier alternative.

Do you observe same irregularities on your setup?
Do you maybe see some clear way to use this info easier than window events?

**Edit**
So far, the only way to work with this info is to also check if node type is `Container`.
In my very limited testing this seems to always select correct node, but it would be nice to get confirmation from `i3` in the form of official documentation for the `fullscreen_mode` parameter.

**Comment by ChristianJacobsen** [comment]
Hi!

So what I'm using this for is this:
```cpp
#include <i3-ipc++/i3_ipc.hpp>

#include <algorithm>
#include <iostream>

bool find_fullscreen(const i3_containers::node& con);
void on_window_event(const i3_containers::window_event& event);
void set_dpms(bool state);

static i3_ipc i3;

int main() {
    i3.on_window_event(on_window_event);

    while(true) {
        i3.handle_next_event();
    }

    return 0;
}

bool find_fullscreen(const i3_containers::node& con) {
    if (con.type == i3_containers::node_type::con && con.fullscreen_mode) {
        return true;
    }

    return std::any_of(
        con.nodes.cbegin(),
        con.nodes.cend(),
        find_fullscreen
    );
}

void on_window_event(const i3_containers::window_event& event) {
    if (event.change == i3_containers::window_change::fullscreen_mode
        || event.change == i3_containers::window_change::close) {
        set_dpms(!find_fullscreen(i3.get_tree()));
    }
}

void set_dpms(bool state) {
    if (state) {
        std::cout << "DPMS On\n";
        i3.execute_commands("exec xset s on");
        i3.execute_commands("exec xset +dpms");
    }
    else {
        std::cout << "DPMS Off\n";
        i3.execute_commands("exec xset s off");
        i3.execute_commands("exec xset -dpms");
    }
}
```

The reason I need to check whether the node/container is in `fullscreen_mode` is because I also have to check for other windows in fullscreen, to make sure I don't turn DPMS on too soon.

If you look [here](https://i3wm.org/docs/ipc.html) in section `4.6. window event`, there is a description for `fullscreen_mode` on the `window_event`. It seems that `fullscreen_mode` is whether that specific window entered or exited fullscreen, but it might not be set otherwise and default to a falsy value with the changes I've made in my PR. Maybe this would be better as an optional value?

Please let me know what you think and then I'll try implemented the feature 'correctly' if it's still a feature you want in the library. :+1:

Have a great day / weekend when it arrives! :smiley:

**Comment by Iskustvo** [comment]
Well, after seeing your example, I definitely want to support it! :smiley:

Regarding the implementation, I've [asked](https://github.com/i3/i3/issues/3780) people from `i3` to provide us with more details, so we can decide how to implement it.

Have a nice weekend(at least what's left of it)! :smiley:

**Comment by Iskustvo** [comment]
Judging by the [response](https://github.com/i3/i3/issues/3780#issuecomment-526923157), I would suggest following implementation:

#### Commit 1
- Added support for "fullscreen_mode" for "i3_ipc::get_tree" request.

1. Create `enum class fullscreen_mode_type : std::uint8_t` with values `no_fullscreen`, `local_fullscreen` and `global_fullscreen` in [i3_containers.hpp](https://github.com/Iskustvo/i3-ipcpp/blob/master/include/i3_containers.hpp).
2. Add the member `fullscreen_mode_type fullscreen_mode` to `i3_containers::node`.
3. Create helper function `i3_containers::fullscreen_mode_type extract_fullscreen_mode` in unnamed namespace in [i3_json_parser-utility.cpp](https://github.com/Iskustvo/i3-ipcpp/blob/master/src/i3_json_parser/i3_json_parser-utility.cpp).
4. In [i3_json_parser-utility.cpp](https://github.com/Iskustvo/i3-ipcpp/blob/master/src/i3_json_parser/i3_json_parser-utility.cpp) update the `i3_containers::node i3_json_parser::extract_tree` by using the function from step **3.** to extract `fullscreen_mode` value into member from step **2.**.

#### Commit 2
- Extended "get_tree.cpp" example to also print out `fullscreen_mode`.

1. Update `print_i3_tree` function to also prints human readable status for `fullscreen_mode` in [get_tree.cpp](https://github.com/Iskustvo/i3-ipcpp/pull/2) example.

#### Commit 3
- Bumped the minor version number.

1. Bump the minor version number like [this](https://github.com/Iskustvo/i3-ipcpp/commit/084703f4ef22ffe6d33c1595258de7a4c7497305).

---

If **anything** is unclear or you think it can be done/named better, please reply here so we can sort it out.

---

In the meantime, I will update `i3`'s documentation for `fullscreen_mode`.

**Comment by ChristianJacobsen** [comment]
Sounds like a plan! I'm on it :+1:
Should be able to get it done today ~~- worst case tomorrow.~~

**Comment by ChristianJacobsen** [comment]
Done :)

File: include/i3_containers.hpp [hunk 1]
```diff
@@ -125,6 +125,16 @@ namespace i3_containers
...
 
+    /**
+     * Enumerator used to describe the fullscreen mode of a window.
+     */
+    enum class fullscreen_mode_type : std::uint8_t
+    {
+        no_fullscreen,    /**< No fullscreen. */
+        local_fullscreen, /**< Local fullscreen (window is fullscreened on its output). */
+        global_fullscreen /**< Global fullscreen (window is globally fullscreened). */
+    };
```
Comment (Iskustvo) [inline include/i3_containers.hpp:136]: Please align the comment ending sequence to be in the same column.

File: src/i3_json_parser/i3_json_parser-utility.cpp [hunk 2]
```diff
@@ -181,6 +181,22 @@ namespace
...
         return window;
     }
+
+    /**
+     * \brief                      Extract the fullscreen_mode of the window from the JSON object.
+     * 
+     * \param [in] a_json_object   JSON object.
+     * 
+     * \return                     Extracted fullscreen_mode.
+     */
```
Comment (Iskustvo) [inline src/i3_json_parser/i3_json_parser-utility.cpp:191]: Please replace `Extract` with `Extracts` and remove trailing whitespaces in this paragraph.

File: src/i3_json_parser/i3_json_parser-utility.cpp [hunk 3]
```diff
@@ -181,6 +181,22 @@ namespace
...
+    i3_containers::fullscreen_mode_type extract_fullscreen_mode(const rapidjson::Value& a_json_object)
+    {
+        assert(a_json_object.IsObject());
+        assert(a_json_object.HasMember("fullscreen_mode") && a_json_object["fullscreen_mode"].IsNumber());
```
Comment (Iskustvo) [inline src/i3_json_parser/i3_json_parser-utility.cpp]: We can be stricter and use `IsUint` instead of `IsNumber`.

File: examples/get_tree.cpp [hunk 4]
```diff
@@ -153,6 +153,21 @@ void print_i3_tree(const i3_containers::node& a_node, const std::string& a_inden
     }
     std::cout << "\n";
 
+    // Print the fullscreen_mode of the window
```
Comment (Iskustvo) [inline examples/get_tree.cpp]: Please add full stop at the end of the sentence.

File: examples/get_tree.cpp [hunk 5]
```diff
@@ -153,6 +153,21 @@ void print_i3_tree(const i3_containers::node& a_node, const std::string& a_inden
...
+    std::string fullscreen_mode;
+    switch(a_node.fullscreen_mode) {
```
Comment (Iskustvo) [inline examples/get_tree.cpp]: Please move curly bracket in the new line.

File: include/i3_containers.hpp [hunk 6]
```diff
@@ -148,6 +158,7 @@ namespace i3_containers
         std::vector<std::uint64_t> focus;                 /**< List of child node IDs in focus order.                 */
         std::vector<node> nodes;                          /**< The tiling child containers of this node.              */
         std::vector<node> floating_nodes;                 /**< The floating child containers of this node.            */
+        fullscreen_mode_type fullscreen_mode;             /**< The fullscreen mode of the container.                  */
```
Comment (Iskustvo) [inline include/i3_containers.hpp]: Please move this member above `nodes` to match the order of the `i3` documentation.

**Review by Iskustvo — CHANGES_REQUESTED** [review]
Great job, thank you! :+1: 
There are few cosmetic changes I would like for you to apply, just to keep the project consistent.
Once you are done, please rearrange commits in agreed format and force-push the branch to make GitHub update this pull request.

If you don't know how to rearrange commits in specified format, without repeating changes manually, just push the review changes ~~as another commit~~ on separate commits(one commit should be a fixup for one of the original commits) and I will show you how to do it afterwards.

Just to clarify the last paragraph with the example:
Review changes for commit `Refactors the fullscreen_mode into ENUM class ` should be in commit `FIXUP - Refactors the fullscreen_mode into ENUM class `.
Review changes for commit `Updates get_tree example with fullscreen_mode` should be in commit `FIXUP - Updates get_tree example with fullscreen_mode`.

**Comment by ChristianJacobsen** [comment]
I think the changes should be to your specification :)

If there's anything else, please don't hesitate! :)

**Comment by Iskustvo** [comment]
They are, thanks! :+1: 
Please, just one more thing :smile:
Can you squash first two commits, since the second one is just a fixup for the first one.

With that change, we will honor the original plan:

> First commit: Add feature.
> Second commit: Update examples.
> Third commit: Update version number.

**Comment by ChristianJacobsen** [comment]
I think that should do it!

**Comment by Iskustvo** [comment]
Thank you for the PR, it's been a pleasure working with you!
Best of luck!

**Comment by ChristianJacobsen** [comment]
I'd love to continue working with you on the project in the future. So holla at me if there's anything I can help with!

Be seeing you!

**Comment by Iskustvo** [comment]
Thank you, I will!

**Pull request merged**